### PR TITLE
isolate gstreamer code & some bugfix

### LINKED ISSUE #153: [feature] Execute command before recording
**User story:** I am a regular screencaster. I want to automate cleaning up my desktop before recording, so that I can quickly and consistently have the same environment every time I record.

## Expected Behavior
* An option to set a pre-recording script.

## Current Behavior
* There is currently no such feature.
* There is a similar feature for post-recording.

## Possible Solution
* Open Preferences
* Under options enable "Execute command before recording"
* Enter command to execute
* When recording is invoked, this command is executed and must exit successfully before recording actually starts
  * If there is a delay enabled, the delay should be timed from when the script ends.
  * If the script fails, recording should not start and a simple notification given ("Recording startup script failed. Please correct the error and try again.").

## Your Environment
* Where did you download the extension? Gnome shell extension website
* Gnome shell version: 3.24.2
* Operating System and version: Ubuntu 17.04
* Display server: X11

### LINKED ISSUE #175: Usage of var/let/const warning
```
gnome-shell[1645]: Some code accessed the property 'ESCoffGIcon' on 
the module 'convenience'. That property was defined with 'let' or 'const' 
inside the module. This was previously supported, but is not correct 
according to the ES6 standard. Any symbols to be exported from a module 
must be defined with 'var'. The property access will work as previously for 
the time being, but please fix your code anyway.
```

Ubuntu 17.10 X-session

### PR DESCRIPTION
**Pull request by iacopodeenosee**
isolated the code related to gstreamer, this should eliminate all bootloop reported so far, added the possibility of running a script before recording as suggested and there are some small bug-fixes.

close #153 
close #175

### CODE REVIEW

File: extension.js [hunk 1]
```diff
@@ -567,33 +603,37 @@ const EasyScreenCast_Indicator = new Lang.Class({
...
+    _doPostCommand :function(){
+        if (Settings.getOption('b', Settings.ACTIVE_POST_CMD_SETTING_KEY)) {
+            Lib.TalkativeLog('-*-execute post command');
 
-                //execute post command
-                this.CtrlExe.Spawn(Cmd);
-            }
-        }
+            //launch cmd after registration
+            var tmpCmd = '/usr/bin/sh -c "' +
```
Comment (tobias47n9e) [inline extension.js]: The 3 "var" in this block could all be "const" because they are never changed.

File: extension.js [hunk 2]
```diff
@@ -519,11 +528,38 @@ const EasyScreenCast_Indicator = new Lang.Class({
...
             //start recording
+            this._doPreCommand();
+        }
+    },
+
+    _doPreCommand: function(){
+        if(Settings.getOption('b', Settings.ACTIVE_PRE_CMD_SETTING_KEY)){
+            Lib.TalkativeLog('-*-execute pre command');
+
+            var PreCmd = Settings.getOption('s',Settings.PRE_CMD_SETTING_KEY);
```
Comment (tobias47n9e) [inline extension.js]: Could be const?

File: extension.js [hunk 3]
```diff
@@ -671,11 +711,12 @@ const EasyScreenCast_Indicator = new Lang.Class({
                     this.indicatorIcon.set_gicon(Lib.ESConGIcon);
                 }
             } else {
-                if (focus === true) {
+                this.indicatorIcon.set_gicon(null);
+/*                if (focus === true) {
```
Comment (tobias47n9e) [inline extension.js]: Remove the commented code?

File: prefs.js [hunk 4]
```diff
@@ -45,7 +45,10 @@ const EasyScreenCastSettingsWidget = new GObject.Class({
         // creates the settings
         Settings.checkSettings();
 
-        this.CtrlWebcam = null;
+        /*
+        this.CtrlWebcam = new UtilWebcam.HelperWebcam();
```
Comment (tobias47n9e) [inline prefs.js]: Commented code.

File: prefs.js [hunk 5]
```diff
@@ -78,13 +81,15 @@ const EasyScreenCastSettingsWidget = new GObject.Class({
             this._initTabQuality(this, builder, Settings.settings);
 
             // setup tab webcam
+            /*
             this._initTabWebcam(this, builder, Settings.settings);
```
Comment (tobias47n9e) [inline prefs.js]: Commented code

File: prefs.js [hunk 6]
```diff
@@ -98,6 +103,11 @@ const EasyScreenCastSettingsWidget = new GObject.Class({
             this._updateRowShortcut(
                 Settings.getOption('as', Settings.SHORTCUT_KEY_SETTING_KEY)[0]);
 
+            //update webcam widget state
+            /*
+            this._updateStateWebcamOptions();
```
Comment (tobias47n9e) [inline prefs.js]: Commented code

File: prefs.js [hunk 7]
```diff
@@ -662,97 +678,115 @@ const EasyScreenCastSettingsWidget = new GObject.Class({
...
-                            if(Ref_buffer_Log.get_line_count() > 0){
-                                let strNOgsp = _('No Gstreamer pipeline found');
+        //implements file resolution preset spinner
+        let Ref_combobox_LogChooser = gtkDB.get_object(
+            'cbt_LogChooser');
+
+        //intercept combobox res changed and update width/height value
+        Ref_combobox_LogChooser.connect('changed',
+            (self) => {
+                var activeLog = self.active;
```
Comment (tobias47n9e) [inline prefs.js]: Could be const?

File: extension.js [hunk 8]
```diff
@@ -59,7 +59,10 @@ const EasyScreenCast_Indicator = new Lang.Class({
         Settings.setOption(Settings.DEVICE_WEBCAM_SETTING_KEY, -1);
 
         this.CtrlAudio = new UtilAudio.MixerAudio();
-        this.CtrlWebcam  = null;
+        /*
+        this.CtrlWebcam = new UtilWebcam.HelperWebcam();
```
[the remark this answers is not in the document]
Comment (iacopodeenosee) [reply extension.js]: the commented code will be valid as soon as I have time to check the problem with the initialization of gstreamer, I would leave it commented.

File: extension.js [hunk 9]
```diff
@@ -124,7 +127,9 @@ const EasyScreenCast_Indicator = new Lang.Class({
         this._addSubMenuAudioRec();
 
         //add sub menu webcam recording
+        /*
```
[the remark this answers is not in the document]
Comment (iacopodeenosee) [reply extension.js]: ok I did it for each commented part

File: extension.js [hunk 10]
```diff
@@ -497,18 +502,22 @@ const EasyScreenCast_Indicator = new Lang.Class({
     _enable: function() {
         //enable key binding
         this._enableKeybindings();
+        //start monitoring inputvideo
+        /*
+        this.CtrlWebcam.startMonitor();
```
[the remark this answers is not in the document]
Comment (iacopodeenosee) [reply extension.js]: as above

File: extension.js [hunk 11]
```diff
@@ -497,18 +502,22 @@ const EasyScreenCast_Indicator = new Lang.Class({
...
         //remove key binding
         this._removeKeybindings();
-
-        if (Indicator.CtrlWebcam !== null) {
-            //stop monitoring inputvideo
-            Indicator.CtrlWebcam.stopMonitor();
-        }
+        //stop monitoring inputvideo
+        /*
+        this.CtrlWebcam.stopMonitor();
```
[the remark this answers is not in the document]
Comment (iacopodeenosee) [reply extension.js]: as above

File: extension.js (hunk 2 above)
Comment (iacopodeenosee) [reply extension.js]: ok, fixed

File: extension.js (hunk 1 above)
Comment (iacopodeenosee) [reply extension.js]: fixed

File: extension.js (hunk 3 above)
Comment (iacopodeenosee) [reply extension.js]: fixed issue: https://github.com/EasyScreenCast/EasyScreenCast/issues/155

File: prefs.js (hunk 4 above)
Comment (iacopodeenosee) [reply prefs.js]: as above

File: prefs.js (hunk 5 above)
Comment (iacopodeenosee) [reply prefs.js]: as above

File: prefs.js (hunk 6 above)
Comment (iacopodeenosee) [reply prefs.js]: as above

File: prefs.js (hunk 7 above)
Comment (iacopodeenosee) [reply prefs.js]: fixed

**Pull request merged**